enhancements/update/eus-upgrades-mvp.md

```diff
@@ -0,0 +1,368 @@
+
+#### OLM - Enforce defined OLM managed Operator compatibility
+
+If a managed Operator defines a maxKubeVersion or maxOCPVersion which would be
+violated by upgrading to the next minor OLM must set Upgradeable=False and
+enumerate the constraint in the condition message.
+
+Note, this is up for debate as to what the behavior should be in absence of a
+defined maximum version.
+
+#### OTA - Optimize & Model Minor Version Upgrade Duration
+
+With support of the PerfScale team the Over The Air team will measure and
+analyze upgrade timings using simulated workloads at a defined capacity and
+load. This analysis will be used to identify optimizations and set expectations
+on a per minor version basis as to the duration of the upgrade.
+
+Due to the variability in workload deployment these measurements will be
+*exclusive of Worker MCP rollout*. The simulated workload is intended only to
+ensure that our measurements more closely mimic real world usage, not upgrades
+of clusters with zero workload.
+
+##### SDN - Canary and Optimize SDN DaemonSet rollout
+
+In 4.6 and 4.7 we identified a number of DaemonSets including dns, mcd, node-tuned,
+node-exporter which were not critical to workload availability which could rollout
+in a more parallel manner. Those operators were changed to utilize maxUnavailable
+of 10% rather than absolute value of 1, on a cluster of 250 nodes that generally
+reduced rollout time from 80 minutes to 10 minutes. The multus, sdn, ovs DaemonSets
+have not been updated but likely could be now that workload critical processes
+have been moved to the host. Before doing this given the risk we should devise
+a method to canary updates to these DaemonSets. Any other options to speed up
+updates to these DaemonSets which are deployed to all Linux nodes should also be
+considered.
+
+Once these new deployment patterns have been vetted in 4.8 or 4.9 these changes
+should be backported through to 4.6.
+
+#### Workloads - Optimize scheduling for rolling upgrades
+
+Currently the MCO cycles nodes in a deterministic but unintentional manner. The
+workload on that host which can be relocated scatters to other hosts according
+to scheduling rules. This has been measured and shown that approximately 70% of
+all reschedulable pods are rescheduled more than necessary if they were
+rescheduled in an optimal manner. Optimal rescheduling should be closer to (1 /
+node_count_per_failure_domain) percentage of pods being rescheduled more than
+once during a rolling restart of a given failure domain.
+
+We should make the MCO topology aware so that it upgrades all of the host in one
+failure domain because anti-affinity scheduling is used in most scenarios. We
+should provide some mechanism to rank upgraded nodes as slightly higher priority
+in scheduler configuration. Perhaps a timestamp of configuration generation that
+the scheduler then biases toward nodes with the greatest timestamps as a last in
```

Comment:
This seems achievable and super-useful.  Currently the MCO annotates nodes with current and target MachineConfig references.  It could also annotate them with a rank or timestamp or some such of the current MachineConfig, and the scheduler could use that rank when breaking ties to increase the chances that drained workloads land on the already-updated nodes.  With anti-affinity rules and all that usual stuff protecting workloads from getting overly wedged into a particular failure domain, even if that means temporarily scheduling some workload on still-to-be-updated nodes.

Reply:
I think using a timestamp, while easy enough, is likely going to result in a problem as its does not indicate whether the nodes have the preferred state. I *think* having the MCO introduce the idea of an "update-epoch" would be clear and allow the scheduler to send workloads based on nodes with the highest epoch. Something like `machineconfiguration.openshift.io/update-epoch: 20` (this would be interesting for telemetry, too).

However, if someone has paused their machine config pools, it would have the effect of preferring to run workloads on unpaused pools (since they would have a higher epoch).